Entity refactor

JWR26 opened the pull request:
This is an attempt to refactor out the entities (see #4) in order to regroup common behaviour accross the entities. It is a large change but I hope it will give a certain modularity to the code so extra features can be added easily if desired.

Firstly, entities now all inherit from a base class `Entity2D`. This holds methods for taking damage, two of which can be overriden by inheriting classes. There are no keywords or ways to identify clearly methods that can be overloaded/are overloads (such as using the `virtual` & `overload` keywords in C#).

`Asteroid` now derrives from `Entity2D` and `entity_destroyed()` is overriden so the `asteroid_destroyed` is emitted upon destruction.

The `Ship2D` class has been created, inheriting from the `Entity2D` class. It contains the common functionality for both health and bouncing bullet powerups. Also, a single method `_fire_projectile` can be called by any inheriting classes. All of the required parameters must be passed in, with only the projectile colour defaulting to white. The size of projectiles is determined by the `projectile_radius` variable, which is exported to allow quick customisation from the editor. One consequence of this base class for ships is that powerups are not exclusive to the player. However, UFO's collecting powerups are currently happy accidents. Logic should be implemented into the UFO script to get them to actively seek the powerups.

Both `Player2D` & `UFO` now inherit from the `Ship2D`, each handling their own physics process and firing of projectiles.

Following the refactor, I decided to change the `Projectile2D` to be a stand alone script (in the same style as the `Asteroid` class). I would be very keen to get feeback on this. Regarding #9, I am aware this is based onthe `projectile.tscn`, and this branch will have impact on that. I doubt it will be game changing though and any conflicts should be easy to manage.

There is an update to the `base_script`, to give a standard template for "what goes where" in the scripts. Combining this with the linter should give a good uniform style to the code. Or, as I have yet to use the linter, it might be useless!

Now I wonder what the best way forward will be. Should this new structure be merged into main before any further changes are made? Does the refactor seem appropriate? Are there any features I have missed or mechanics that have been forgotten? 

One final thing, I think adding some variety could become quite easy. Perhaps there could be some alternatives added to both the asteroids and the UFO to give a little more variety to the gameplay. However, I am conscious that [scope creep](https://en.wikipedia.org/wiki/Scope_creep) must be avoided.

JWR26 committed:
start of refactor

Created a base entity class, with asteroid and ship deriving from this. Ship is inherited by player and UFO.

Also refactored projectiles into a standalone script the same as the asteroid class.

There are lots of bugs present and a loss of functionality in comparison to the main branch.

JWR26 committed:
bouncing bullets correctly implemented

Ship2D's now implement bouncing bullets correctly, allowing both the player and the UFO to have this ability. Projectile2D class has been refined and now can be created from code.

JWR26 committed:
restoring points

points are now scored. There is a problem when the player collides with a UFO as the constant contact removes all health

dslh commented on entities/projectile.gd:
```diff
@@ -38,6 +35,9 @@ func _on_RigidBody2D_body_entered(body: Node) -> void:
 		body.take_damage()
 		queue_free()
+		
+	if body is Entity2D:
+		body.take_damage()
```
I like this, I think groups are a useful tool but in this case we're really just interested in whether the object has a certain method so checking its type makes sense.

Bear in mind that GDscript supports [duck typing](https://godotengine.org/qa/116004/how-to-check-have-the-node-a-custom-function), it can also be useful for this sort of thing, in cases where it's not practical to have everything extend the same base class.

I guess this `if` block should make a call to `queue_free()`. And it should be ok to get rid of the `if body.is_in_group` blocks in this function, right?

dslh commented on entities/entity_asteroid.gd:
```diff
@@ -110,9 +130,8 @@ func compute_mass(outline: PoolVector2Array) -> void:
-	hp -= value
+func entity_destroyed() -> void:
+	EventBus.emit_signal("asteroid_destroyed", 10)
+	queue_free()
```
Prefer to call the parent method, rather than reimplementing it in the override. It reduces the likelihood of bugs when changes are made to the parent method.

```suggestion
	.entity_destroyed()
```

JWR26 replied:
Understood!

dslh commented on entities/entity_ship.gd:
```diff
@@ -0,0 +1,66 @@
+
+
+# --- SetGet functions ---
+func set_bouncing_bullets_active(value: bool = false) -> void:
```
Default value of `false` could be confusing, I think if I saw a line of code that read `set_bouncing_bullets_active()` I would expect it to set `bouncing_bullets_active` to `true`.

JWR26 replied:
Agreed - I have removed the default argument to avoid any confusion

dslh commented on entities/projectile2d.gd:
```diff
@@ -0,0 +1,135 @@
+func set_drawn_shape(radius: int, color: Color) -> void:
+	var polygon = Polygon2D.new()
+	# particles will be a circle
+	var outline: PoolVector2Array = []
+	outline.resize(POINTS)
+	for i in POINTS:
+		var angle: float = PI * 2 * i / POINTS
+		outline[i] = Vector2.UP.rotated(angle) * radius
```
Since the shape is always the same it might be possible to optimise things a bit by generating it once and reusing it with a scale modifier for each instance.

Not that we're having performance issues in this little project, but it could be a fun programming exercise.

dslh commented on entities/entity_ship.gd:
```diff
@@ -0,0 +1,66 @@
+	if max_change < value:
+		take_damage(-max_change)
+	else:
+		take_damage(-value)
```
Implementation could be:

```suggestion
func apply_health_powerup(value: int = 1) -> void:
	# HP can not go over MAX_HP
	# negate the damage to cause HP to increase
	take_damage(-min(MAX_HP - hp, value))
```

I would be tempted to shorten the name to `heal` 😄

JWR26 replied:
I like it a lot!

dslh reviewed (approved):
Great work! This seems like a nice, sensible structure to build on. The `damage_taken` "hook" is a nice pattern, in a larger project you could end up getting good mileage out of this approach.

I think gdlint forces you to put variables, methods and so on in the recommended order, so we might be able to live without the template.

A next step could be to encapsulate weapons/powerups. For example we might have a `Weapon` class with a `fire_projectile` method, and `StandardGun` and `BouncingBulletGun` subclasses. When the related powerup gets collected it calls `ship.set_weapon(BouncingBulletGun.new())`, and then `Player2D._shoot()` calls `weapon.fire_projectile()` twice and `$CannonTimer.start(weapon.get_cooldown())`. Ok it's a bit of a vague plan but the point is that it will be easier to add more weapons and power-ups if we can keep the code for each one in its own file.

But that sort of thing wouldn't be possible without the work you've just done. 😄 I think we should merge this PR and keep iterating from there.

JWR26 committed:
Update entities/entity_asteroid.gd

call of parent method to avoid future bugs

Co-authored-by: Doug Hammond <d.lakehammond@gmail.com>

JWR26 committed:
Update entity_ship.gd

removed default value for set_bouncing_bullets_active()

JWR26 committed:
Update entities/entity_ship.gd

tidying up health powerup function

Co-authored-by: Doug Hammond <d.lakehammond@gmail.com>

JWR26 committed:
comment integration and addition of strafing

Integration of comments on PR10 and the addition of strafing.

JWR26 commented:
I have integrated all comments and also added strafing as suggested in the google doc!

I have also added a "globals.gd" file that heavily repeated functionality can get squished into. I'd be happy for anyone to sort this out or give it correct structure.

I will create an issue for the breaking out of weapons into their own class. This could make it easier to do a feature expansion once the ground is solid!

A note on gameplay, its already getting quite immersive and things start tying in well. I will update the google document to give a bit more structure to the project, which will hopefully make the development easier from here on out.

If you're OK with my changes, I'd like to merge this branch into main and get at least the test rig setup from #9 merged so some more features can be developed!

dslh commented:
Nice, I look forward to giving strafing a go when I get a chance 😄 

globals.gd is a good place to start. If it ends up being a very big file it could be broken up into smaller files with more specific names but there's no need to overthink it for now.

Go ahead, I'd be very happy to see this merged. I'll make sure #9 is free of conflicts.
